js/multimodaltextbox/shared/MultimodalTextbox.svelte

```diff
@@ -0,0 +1,484 @@
+
+	async function handle_copy(): Promise<void> {
+		if ("clipboard" in navigator) {
+			const text = value.find((item) => item && "text" in item);
+			if (text && "text" in text && typeof text["text"] === "string") {
+				await navigator.clipboard.writeText(text["text"]);
+				copy_feedback();
+			}
+		}
+	}
+
+	function copy_feedback(): void {
+		copied = true;
+		if (timer) clearTimeout(timer);
+		timer = setTimeout(() => {
+			copied = false;
+		}, 1000);
+	}
+
+	function handle_select(event: Event): void {
+		const target: HTMLTextAreaElement | HTMLInputElement = event.target as
+			| HTMLTextAreaElement
+			| HTMLInputElement;
+		const text = target.value;
+		const index: [number, number] = [
+			target.selectionStart as number,
+			target.selectionEnd as number
+		];
+		dispatch("select", { value: text.substring(...index), index: index });
+	}
+
+	async function handle_keypress(e: KeyboardEvent): Promise<void> {
+		await tick();
+		if (e.key === "Enter" && e.shiftKey && lines > 1) {
+			e.preventDefault();
+			value.push({ type: "text", text: text });
+			dispatch("submit", value);
+			text = "";
+			await tick();
+			value = [];
+		} else if (
+			e.key === "Enter" &&
+			!e.shiftKey &&
+			lines === 1 &&
+			max_lines >= 1
+		) {
+			e.preventDefault();
+			value.push({ type: "text", text: text });
+			dispatch("submit", value);
+			text = "";
+			await tick();
+			value = [];
+		}
+	}
+
+	function handle_scroll(event: Event): void {
+		const target = event.target as HTMLElement;
+		const current_scroll_top = target.scrollTop;
+		if (current_scroll_top < previous_scroll_top) {
+			user_has_scrolled_up = true;
+		}
+		previous_scroll_top = current_scroll_top;
+
+		const max_scroll_top = target.scrollHeight - target.clientHeight;
+		const user_has_scrolled_to_bottom = current_scroll_top >= max_scroll_top;
+		if (user_has_scrolled_to_bottom) {
+			user_has_scrolled_up = false;
+		}
+	}
+
+	async function resize(
+		event: Event | { target: HTMLTextAreaElement | HTMLInputElement }
+	): Promise<void> {
+		await tick();
+		if (lines === max_lines) return;
+
+		let max =
+			max_lines === undefined
+				? false
+				: max_lines === undefined // default
+				? 21 * 11
+				: 21 * (max_lines + 1);
+		let min = 21 * (lines + 1);
+
+		const target = event.target as HTMLTextAreaElement;
+		target.style.height = "1px";
+
+		let scroll_height;
+		if (max && target.scrollHeight > max) {
+			scroll_height = max;
+		} else if (target.scrollHeight < min) {
+			scroll_height = min;
+		} else {
+			scroll_height = target.scrollHeight;
+		}
+
+		target.style.height = `${scroll_height}px`;
+	}
+
+	function text_area_resize(
+		_el: HTMLTextAreaElement,
+		_value: string
+	): any | undefined {
+		if (lines === max_lines) return;
+		_el.style.overflowY = "scroll";
+		_el.addEventListener("input", resize);
+
+		if (!_value.trim()) return;
+		resize({ target: _el });
+
+		return {
+			destroy: () => _el.removeEventListener("input", resize)
+		};
+	}
```

Comment:
Lets move the action and event handler into a different file to make this more readable.

We'll need to tweak it a little:

- The `_value` here is a string but this can be an object, allowing us to pass in whatever values we need here. So lets change that to accept `lines`, `max_lines` and we will also need to pass those into the event handler. We can change the event handler to a normal function that take in an element, lines and max_lines and adjust the usage:

```js
resize(_el, lines, max_lines);

// and 

_el.removeEventListener("input", (e: Event) => resize(e.target, lines, max_lines))
```